src/sgame/sg_bot_nav.cpp

```diff
@@ -27,14 +27,123 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 #include "botlib/bot_types.h"
 #include "botlib/bot_api.h"
 #include "shared/bot_nav_shared.h"
+#include "shared/navgen/navgen.h"
 
 #include "shared/bg_local.h"
 
 #include <glm/geometric.hpp>
 #include <glm/gtx/norm.hpp>
 
 //tells if all navmeshes loaded successfully
-bool navMeshLoaded = false;
+// Only G_BotNavInit and G_BotNavCleanup should set it
+navMeshStatus_t navMeshLoaded = navMeshStatus_t::UNINITIALIZED;
+
+/*
+===========================
+Navigation Mesh Generation
+===========================
+*/
+
+// blocks the main thread!
+void G_BlockingGenerateNavmesh( std::bitset<PCL_NUM_CLASSES> classes )
+{
+	std::string mapName = Cvar::GetValue( "mapname" );
+	NavmeshGenerator navgen;
+
+	for ( int i = PCL_NONE; ++i < PCL_NUM_CLASSES; )
+	{
+		if ( !classes[ i ] )
+		{
+			continue;
+		}
+
+		navgen.Init( mapName );
+		navgen.StartGeneration( Util::enum_cast<class_t>( i ) );
+
+		while ( !navgen.Step() )
+		{
+			// ping the engine with a useless message so that it does not think the sgame VM has hung
+			Cvar::GetValue( "x" );
```

Comment:
At some point we may want to implement a specific ping for that, but that's not blocking this PR.